packages/tanstack-start/src/Transport.ts

```diff
@@ -0,0 +1,153 @@
+import type {
+  DataTransportContext,
+  QueryEvent,
+} from "@apollo/client-react-streaming";
+
+import { createSerializationAdapter } from "@tanstack/router-core";
+import type { Observable } from "rxjs";
+import React from "react";
+
+interface ValueEvent {
+  type: "value";
+  value: unknown;
+  id: string;
+}
+
+type Transported = ReadableStream<QueryEvent | ValueEvent>;
+
+type DataTransportAbstraction =
+  typeof DataTransportContext extends React.Context<infer T>
+    ? NonNullable<T>
+    : never;
+
+export type TransportSerializationAdapter =
+  typeof transportSerializationAdapter;
+export const transportSerializationAdapter = createSerializationAdapter<
+  ServerTransport | ClientTransport,
+  Transported
+>({
+  key: "apollo-transport",
+  test(value): value is ServerTransport {
+    return value instanceof ServerTransport;
+  },
+  toSerializable(value) {
+    // TS is a bit too strict about serializability here - some values are just `unknown`, but definitely serializable
+    return (value as ServerTransport).stream satisfies Transported as any;
+  },
+  fromSerializable(value) {
+    return new ClientTransport(value);
+  },
+});
+
+export class ServerTransport implements DataTransportAbstraction {
+  stream: Transported;
+  private controller!: ReadableStreamDefaultController<QueryEvent | ValueEvent>;
+  private ongoingStreams = new Set<Extract<QueryEvent, { type: "started" }>>();
+
+  constructor() {
+    this.stream = new ReadableStream<QueryEvent | ValueEvent>({
+      start: (controller) => {
+        this.controller = controller;
+      },
+    });
+  }
+
+  private shouldClose = false;
+  private closed = false;
+  closeOnceFinished() {
+    this.shouldClose = true;
+    if (this.ongoingStreams.size === 0 && !this.closed) {
+      this.controller.close();
+      this.closed = true;
+    }
+  }
+
+  dispatchRequestStarted = ({
+    event,
+    observable,
+  }: {
+    event: Extract<QueryEvent, { type: "started" }>;
+    observable: Observable<Exclude<QueryEvent, { type: "started" }>>;
+  }): void => {
+    this.controller.enqueue(event);
+    this.ongoingStreams.add(event);
+    const finalize = () => {
+      this.ongoingStreams.delete(event);
+      if (this.shouldClose && this.ongoingStreams.size === 0 && !this.closed) {
+        this.controller.close();
+        this.closed = true;
+      }
```

Comment:
You might consider extracting a private `close` function that moves this logic in there. I see the same thing is used on lines 59-62 so this should probably live together (in case you need to make changes later, you don't end up forgetting to modify both).

```ts
private close() {
  if (this.ongoingStreams.size === 0 && !this.closed) {
    this.controller.close();
    this.closed = true;
  }
}
```

Reply:
3fbf55462d33ee379a7de418939a76f77d9e5087